tedana/metrics/external_regressor_fits.py

```diff
@@ -0,0 +1,751 @@
+            )
+
+        if len(external_regressors.index) != n_time:
+            err_msg += (
+                f"External Regressors have len(external_regressors.index) timepoints\n"
+                f"while fMRI data have {n_time} timepoints"
+            )
+
+        if err_msg:
+            raise RegressError(err_msg)
+
+
+def fit_regressors(comptable, external_regressors, external_regressor_dict, mixing):
+    """
+    Fit regressors to the mixing matrix.
+
+    Uses correlation or F statistics in a linear model depending on the calc_stats
+    value in external_regressor_dict
+
+    Parameters
+    ----------
+    comptable : (C x X) :obj:`pandas.DataFrame`
+        Component metric table. One row for each component, with a column for
+        each metric. The index is the component number.
+    external_regressors : :obj:`pandas.DataFrame`
+        Each column is a labelled regressor and the number of rows should
+        match the number of timepoints in the fMRI time series
+    external_regressor_dict : :obj:`dict`
+        Information describing the external regressors and
+        method to use for fitting and statistical tests
+    mixing : (T x C) array_like
+        Mixing matrix for converting input data to component space, where `C`
+        is components and `T` is the same as in `data_cat`
+
+    Returns
+    -------
+    comptable : (C x X) :obj:`pandas.DataFrame`
+        Component metric table. Same as inputted, with additional columns
+        for metrics related to fitting the external regressors
+    """
+    n_time = mixing.shape[0]
+
+    # If the order of polynomial detrending is specified, then pass to make_detrend_regressors
+    # otherwise the function sets a detrending polynomial order
+    if external_regressor_dict["detrend"] is True:
+        detrend_regressors = make_detrend_regressors(n_time, polort=None)
+    elif (
+        isinstance(external_regressor_dict["detrend"], int)
+        and external_regressor_dict["detrend"] > 0
+    ):
+        detrend_regressors = make_detrend_regressors(
+            n_time, polort=external_regressor_dict["detrend"]
+        )
+    else:
+        LGR.warning("External regressor fitting applied without detrending fMRI time series")
+
+    if external_regressor_dict["calc_stats"].lower() == "corr":
+        if "detrend_regressors" not in locals():
+            # set detrend regressors to None if it doesn't exist
+            detrend_regressors = None
+        comptable = correlate_regressors(
+            comptable, external_regressors, mixing, detrend_regressors=detrend_regressors
+        )
+    elif external_regressor_dict["calc_stats"].lower() == "f":
+        # external_regressors = pd.concat([external_regressors, detrend_regressors])
+        comptable = fit_mixing_to_regressors(
+            comptable, external_regressors, external_regressor_dict, mixing, detrend_regressors
+        )
+
+    return comptable
+
+
+def make_detrend_regressors(n_time, polort=None):
```

Comment:
Aren't some of the largest-variance components typically characterized by drifts? Are there some cases where we'd want to detrend regressors, but other cases where we wouldn't?

Reply:
Thinking about this, I think I'm doing something non-ideal here and in my earlier non-tedana version of this analysis. We need to model slow drifts because, they can be a large source of variance and our external regressors will less often model a significant portion of the variance if we don't model the drift. That said, if we're calculating fits based on the full F statistic model then tedana is detrending data by default. This isn't good because a benefit of tedana is it can potentially separate and retain T2*-weight slow drifts in learning or pharmacology studies.

I think the better approach is to output a partial model with all nuissance regressors except the slow drifts and use that in the decision tree. Based on feedback from Logan & Gang Chen, I might also give an option to model task regressors so decision trees might skew towards keeping them.

Reply:
I was confused in my above comment. The Full F test is comparing a model with all the external regressors and the detrending regressors to just the detrending regressors. That is, it's detrending the data and seeing if the remaining regressors model the signal above and beyond what is detrended. As such, this is what I think we want. A component won't be flagged and removed because of a large signal drift unless that componend is flagged for rejection based on kappa/rho criteria. If a drift component is T2* weighted, tedana should not remove it (but a user can remove it separately)